File: Python/QuickSort.py
def Swap(A, i, j):
	t = A[i]
	A[i] = A[j]
	A[j] = t

def Partition(A, l, r):
	p = A[l]
	i = l + 1
	for j in range(l + 1, r + 1):
		if A[j] < p:
			Swap(A, i, j)
			i = i + 1
	Swap(A, l, i - 1)
	return i - 1

def QuickSortRecursive(A, l, r):
	if l >= r:
		return
	i = Partition(A, l, r)
	QuickSortRecursive(A, l, i - 1)
	QuickSortRecursive(A, i + 1, r)

def QuickSort(array):
	QuickSortRecursive(array, 0, len(array) - 1)

File: Java/src/QuickSort.java
import java.util.Random;

public class QuickSort implements Sort
{
	private Random rand = new Random();
	
	public void execute(int[] array)
	{
		executeRec(array, 0, array.length - 1);
	}
	
	private void executeRec(int[] array, int l, int r)
	{
		if (l >= r) return;
		int p = partition(array, l, r);
		executeRec(array, l, p - 1);
		executeRec(array, p + 1, r);
	}
	
	private int partition(int[] array, int l, int r)
	{
		swap(array, l, rand.nextInt(r - l + 1) + l);
		int p = array[l];
		int j = l + 1;
		for (int i = l + 1; i <= r; ++i)
			if (array[i] < p)
				swap(array, i, j++);
		swap(array, l, j - 1);
		return j - 1;
	}
	
	private void swap(int[] array, int i, int j)
	{
		int t = array[i];
		array[i] = array[j];
		array[j] = t;
	}
}
